### ISSUE TITLE
Android: Corrupt images rendered in SVG

### ISSUE THREAD

**Issue by silverspace** [magicismight/react-native-svg#80]
I have been debugging two issues with RNSVGImageShadowNode: 
  1.  My SVG image is often 'corrupt', with a smaller mirror copy of the image inside of the larger image. (See screenshot). 
  2. I am often receiving an error from Canvas about "Underflow in restoreToCount - more restores than saves". 

I believe that both of these issues are actually caused by RNSVGImageShadowNode::loadBitmap(). 
More specifically:
  1. onNewResultImpl() is directly saving a reference to the bitmap object, which is not allowed in BaseBitmapDataSubscriber:  "The bitmap provided to this method is only guaranteed to be around for the lifespan of the method. The framework will free the bitmap's memory after this method has completed." At the very least, this should be   mBitmap = bitmap.copy(bitmap.getConfig(), false);

  2. onNewResultImpl() is calling node.drawChildren(canvas, paint), which might happen in a separate thread. It looks like this code is assuming that the same thread will be used in draw() and onNewResultImpl since it's using CallerThreadExecutor.getInstance(), but that is not the case. (I observed different thread ids in each method). 
  * This issue is what is causing the Canvas errors with mismatched calls to save and restore, since the threads are concurrently modifying the Canvas element. 

(Tested on react-native-0.30.0 with react-native-svg-3.1.1)

`
java.lang.IllegalArgumentException: Underflow in restoreToCount - more restores than saves
at android.graphics.Canvas.native_restoreToCount(Native Method)
at android.graphics.Canvas.restoreToCount(Canvas.java:566)
at com.horcrux.svg.RNSVGVirtualNode.restoreCanvas(RNSVGVirtualNode.java:103)
at com.horcrux.svg.RNSVGTextShadowNode.draw(RNSVGTextShadowNode.java:94)
at com.horcrux.svg.RNSVGGroupShadowNode.draw(RNSVGGroupShadowNode.java:36)
at com.horcrux.svg.RNSVGGroupShadowNode.draw(RNSVGGroupShadowNode.java:36)
at com.horcrux.svg.RNSVGGroupShadowNode.draw(RNSVGGroupShadowNode.java:36)
at com.horcrux.svg.RNSVGSvgViewShadowNode.drawChildren(RNSVGSvgViewShadowNode.java:59)
at com.horcrux.svg.RNSVGSvgViewShadowNode.drawOutput(RNSVGSvgViewShadowNode.java:50)
at com.horcrux.svg.RNSVGSvgViewShadowNode.onCollectExtraUpdates(RNSVGSvgViewShadowNode.java:39)
at com.facebook.react.uimanager.ReactShadowNode.dispatchUpdates(ReactShadowNode.java:200)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:758)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:748)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:748)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:748)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:748)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:748)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:748)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:748)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:748)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:748)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:748)
at com.facebook.react.uimanager.UIImplementation.applyUpdatesRecursive(UIImplementation.java:748)
at com.facebook.react.uimanager.UIImplementation.updateViewHierarchy(UIImplementation.java:506)
at com.facebook.react.uimanager.UIImplementation.dispatchViewUpdates(UIImplementation.java:494)
at com.facebook.react.uimanager.UIManagerModule.onBatchComplete(UIManagerModule.java:457)
at com.facebook.react.bridge.NativeModuleRegistry.onBatchComplete(NativeModuleRegistry.java:137)
at com.facebook.react.bridge.CatalystInstanceImpl$NativeModulesReactCallback.onBatchComplete(CatalystInstanceImpl.java:453)
at com.facebook.react.bridge.queue.NativeRunnableDeprecated.run(Native Method)
at android.os.Handler.handleCallback(Handler.java:746)
at android.os.Handler.dispatchMessage(Handler.java:95)
at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:31)
at android.os.Looper.loop(Looper.java:148)
at com.facebook.react.bridge.queue.MessageQueueThreadImpl$3.run(MessageQueueThreadImpl.java:193)
at java.lang.Thread.run(Thread.java:818)
`

**Comment by silverspace** [comment]
<img width="117" alt="corrupt_svg_render" src="https://cloud.githubusercontent.com/assets/15318933/17075440/3766ff5a-5048-11e6-99e2-2f16f7d858c8.png">

**Comment by silverspace** [comment]
I've created a PR with a fix for this issue:  https://github.com/magicismight/react-native-svg/pull/86

This PR has the following fixes:
1. Fetching the bitmap from the Fresco image cache instead of a local bitmap, since the current local bitmap is a dangerous reference to a bitmap that should not be accessed outside of onNewResultImpl(). (This dangerous reference may also be causing memory leaks). 

2. Synchronizing calls to drawChildren(), since multiple threads can call it simultaneously and render to the canvas object.

Verified that this PR fixes my double image rendering and canvas underflow issues with RN-0.30.0 and my patch on top of RNSVG-3.1.1.
